src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -263,23 +318,24 @@ steps m@Merge {..} requestedSteps = assertStepsInvariant <$> do
     -- to satisfy this precondition when it calls @steps@, so for now we do the
     -- check.
     readMutVar mergeState >>= \case
-      Merging -> case mergeType of
-        MergeMidLevel  -> doStepsLevel m requestedSteps
-        MergeLastLevel -> doStepsLevel m requestedSteps
-        MergeUnion     -> doStepsUnion m requestedSteps
+      Merging -> case toMergeType mergeType of
+        MergeTypeMidLevel  -> doStepsLevel m requestedSteps
+        MergeTypeLastLevel -> doStepsLevel m requestedSteps
+        MergeTypeUnion     -> doStepsUnion m requestedSteps
```

Comment:
This use of `toMergeType` is what prevents the `SPECIALISE` above from having no unresolved class dicts.

Reply:
Comment turns out to be made redundant by later patches.